packages/firestore/src/core/firestore_client.ts

```diff
@@ -665,6 +656,9 @@ function readDocumentViaSnapshotListener(
 ): Promise<void> {
   const wrappedObserver = new AsyncObserver({
     next: (snap: ViewSnapshot) => {
+      // No further messages should be processed.
+      wrappedObserver.mute();
```

Comment:
This actually relates to next part of code:

```
      // Remove query first before passing event to user to avoid
      // user actions affecting the now stale query.
      asyncQueue.enqueueAndForget(() =>
        eventManagerUnlisten(eventManager, listener)
      );
```

When debugging failing tests, I ran across this.

I don't recall whether `mute()` was strictly required to fix the test that was failing, but looking at the code, you should see this is a race condition. If the AsyncQueue emits more snapshots before the unlisten is processed, then "Remove query first" doesn't really help. We need to `mute()` the observer to avoid multiple calls to result callback.

This pattern exists for other listeners, but for some reason was not applied here.

I moved code down to be part of unlisten.

If you still have concerns, I can remove change, and have a separate PR for this.